### PR TITLE
Ltac2 preterm antiquotation `$preterm:x`

### LINKED ISSUE #13977: `uconstr:` for Ltac2?
When constructing a large term in Ltac1/Ltac2, it is important for performance that we don't re-typecheck it at each step, but only at the end. In Ltac1, we can use `uconstr:()` for this purpose:

```coq
Require Import ZArith. Open Scope Z_scope.

Ltac z2nat_constr x :=
  let b := eval cbv in (Z.leb x 0) in
  lazymatch b with
  | true => constr:(O)
  | false => let pred := eval cbv in (Z.pred x) in
             let r := z2nat_constr pred in
             constr:(S r)
  end.

Ltac z2nat_uconstr x :=
  let b := eval cbv in (Z.leb x 0) in
  lazymatch b with
  | true => uconstr:(O)
  | false => let pred := eval cbv in (Z.pred x) in
             let r := z2nat_uconstr pred in
             uconstr:(S r)
  end.

Goal False.
  Time let r := z2nat_uconstr 10000 in idtac r. (* 0.748 secs *)
  Time let r := z2nat_constr  10000 in idtac r. (* 2.453 secs *)

  Time let r := z2nat_uconstr 20000 in idtac r. (* 1.236 secs *)
  Time let r := z2nat_constr  20000 in idtac r. (* 9.034 secs *)
Abort.
```

However, it seems that `uconstr:()` is not available in Ltac2, and I have to resort to the `Unsafe` module to achieve the desired speedup:

```coq
Require Import Ltac2.Ltac2.

Ltac2 rec z2nat_constr x :=
  let b := eval cbv in (Z.leb $x 0) in
  lazy_match! b with
  | true => constr:(O)
  | false => let pred := eval cbv in (Z.pred $x) in
             let r := z2nat_constr pred in
             constr:(S $r)
  end.

Ltac2 rec z2nat_open_constr x :=
  let b := eval cbv in (Z.leb $x 0) in
  lazy_match! b with
  | true => open_constr:(O)
  | false => let pred := eval cbv in (Z.pred $x) in
             let r := z2nat_open_constr pred in
             open_constr:(S $r)
  end.

(* array literals would be nice... https://github.com/coq/coq/issues/13976 *)
Ltac2 singleton(x: 'a) := Array.make 1 x.

Ltac2 mk_app(a: constr)(b: constr) :=
  Constr.Unsafe.make (Constr.Unsafe.App a (singleton b)).

Ltac2 rec z2nat_mk_app x :=
  let b := eval cbv in (Z.leb $x 0) in
  lazy_match! b with
  | true => Env.instantiate reference:(O)
  | false => let pred := eval cbv in (Z.pred $x) in
             mk_app (Env.instantiate reference:(S)) (z2nat_mk_app pred)
  end.

Time Ltac2 Eval Message.print (Message.of_constr (z2nat_constr      constr:(10000))). (* 2.762 secs *)
Time Ltac2 Eval Message.print (Message.of_constr (z2nat_open_constr constr:(10000))). (* 3.384 secs *)
Time Ltac2 Eval Message.print (Message.of_constr (z2nat_mk_app      constr:(10000))). (* 0.484 secs *)

Time Ltac2 Eval Message.print (Message.of_constr (z2nat_constr      constr:(20000))). (* 23.351 secs *)
Time Ltac2 Eval Message.print (Message.of_constr (z2nat_open_constr constr:(20000))). (* 22.618 secs *)
Time Ltac2 Eval Message.print (Message.of_constr (z2nat_mk_app      constr:(20000))). (* 0.972 secs *)
```

Therefore, feature request: Please add `uconstr:()` to Ltac2.

Coq version: master (fcfeb5bc45febe1a05f44a0a77b43be6b6905f35)

### PR DESCRIPTION
**Pull request by SkySkimmer**
Using syntax `$$x`

Close https://github.com/coq/coq/issues/13977

Overlays:
- https://github.com/ejgallego/coq-serapi/pull/337

### CODE REVIEW

**Comment by ppedrot** [comment]
I don't like the proliferation of built-in syntax for terms in Ltac2. We should be able to derive it from ltac2 quotations + proper functions to pretype a term. IIUC that's already possible today after #16740?

**Comment by SkySkimmer** [comment]
`ltac2:()` has a significant cost because it needs to turn the env and expected type into a fresh evar

**Comment by ppedrot** [comment]
But then it calls for a more principled approach where we have to provide this kind of expressive antiquotation that prevents the pretyper to redo the work. Just that it's general and needs not be either a constr or a uconstr.

File: doc/sphinx/proof-engine/ltac2.rst [hunk 1]
```diff
@@ -777,6 +777,12 @@ In a Coq term, writing :g:`$x` is semantically equivalent to
 :g:`ltac2:(Control.refine (fun () => x))`, up to re-typechecking. It allows to
 insert in a concise way an Ltac2 variable of type :n:`constr` into a Coq term.
 
+Similarly variables of type `preterm` have an antiquotation
+
+.. prodn:: term += $$@lident
```
Comment (jfehrle) [inline doc/sphinx/proof-engine/ltac2.rst:782]: This production is at LEVEL 200 of `term`?  Same question for `$@lident` a few lines above.  Please update if it's not at that level.

**Comment by SkySkimmer** [comment]
@ppedrot 

I changed the syntax to `$preterm:x` and added `$constr:x` alias of `$x`

File: plugins/ltac2/g_ltac2.mlg [hunk 2]
```diff
@@ -65,6 +65,12 @@ let test_dollar_ident =
     lk_kw "$" >> lk_ident >> check_no_space
   end
 
+let test_dollar_ident_colon_ident =
+  let open Pcoq.Lookahead in
+  to_entry "test_dollar_ident_colon_ident" begin
+    lk_kw "$" >> lk_ident >> lk_kw ":" >> lk_ident >> check_no_space
```
Comment (ppedrot) [inline plugins/ltac2/g_ltac2.mlg:71]: TODO: check this is the correct no-space lookahead.

**Comment by SkySkimmer** [comment]
@coqbot run full ci

**Review by ppedrot — APPROVED** [review]

**Pull request merged**